{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A3.  Shotgun metagenomics - MAG recovery and Functional Prediction \n",
    "\n",
    "**Author: Anna Toidze**\n",
    "\n",
    "In this notebook, I focus on analyzing shotgun metagenomic datasets. I have chosen 10 samples of 5 random patients present in Cohort 1 and Cohort 2 - namely, 'SRR23801859',  'SRR23801860',  'SRR23801866',  'SRR23801898',  'SRR23801908',  'SRR23801935',  'SRR23801941',  'SRR23801944',  'SRR23801959',  'SRR23801999'. Due to computational complexity, all of the following commands have been run the cluster (therefore, different usage of $data_dir than usual - commands run in the same directory, hence, `.`).  \n",
    "\n",
    "\n",
    "**Exercise overview:**<br>\n",
    "[0. Setup](#setup)<br>\n",
    "[1. Sequence preprocessing](#seq)<br>\n",
    "[2. Shotgun metagenome analysis](#moshpit)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[2.1 Genome assembly](#assembly)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[2.2 Read mapping](#mapping)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[2.3 Contig binning](#binning)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[2.4 MAG dereplication](#dereplication)<br>\n",
    "[3. MAG abundance estimation](#moshpit)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[3.1 MAG indexing](#indexing)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[3.2 Read mapping](#mapping)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[3.3 MAG length calculation](#feature-length)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[3.4 Abundance estimation](#abundance)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[3.5 Taxonomic classification of MAGs](#mag_classification)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[3.6 Taxonomic classification of reads](#read_classification)\n",
    "[4. Functional annotation](#annotation)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[4.1 Ortholog search with Diamond](#diamond)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[4.2 Ortholog annotation](#eggnog)<br>\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;[4.3 Annotation extraction](#extraction)<br>\n",
    "[5. Troubleshooting](#trouble)<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='setup'></a>\n",
    "\n",
    "## 0. Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import qiime2 as q2\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "data_dir = 'w9_data'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assigning variables throughout the notebook\n",
    "\n",
    "# location of this week's data and all the results produced by this notebook\n",
    "# - this should be a path relative to your working directory\n",
    "raw_data_dir = \"../data/raw\"\n",
    "data_dir = \"../data/processed\"\n",
    "vis_dir  = \"../results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# setting seaborn formatting style\n",
    "sns.set_style(\"white\")\n",
    "sns.set_theme(style=\"ticks\", palette=\"pastel\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='seq'></a>\n",
    "## 1. Sequence preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Stool_Consistency</th>\n",
       "      <th>Patient_Sex</th>\n",
       "      <th>Sample_Day</th>\n",
       "      <th>Recovery_Day</th>\n",
       "      <th>Cohort_Number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EG2580</td>\n",
       "      <td>P042</td>\n",
       "      <td>liquid</td>\n",
       "      <td>F</td>\n",
       "      <td>13</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EG2559</td>\n",
       "      <td>P043</td>\n",
       "      <td>liquid</td>\n",
       "      <td>M</td>\n",
       "      <td>15</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EG2537</td>\n",
       "      <td>P042</td>\n",
       "      <td>liquid</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EG2518</td>\n",
       "      <td>P043</td>\n",
       "      <td>liquid</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EG2490</td>\n",
       "      <td>P030</td>\n",
       "      <td>formed</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample_Name Patient_ID Stool_Consistency Patient_Sex  Sample_Day  \\\n",
       "0      EG2580       P042            liquid           F          13   \n",
       "1      EG2559       P043            liquid           M          15   \n",
       "2      EG2537       P042            liquid           F           0   \n",
       "3      EG2518       P043            liquid           M           0   \n",
       "4      EG2490       P030            formed           F           0   \n",
       "\n",
       "   Recovery_Day  Cohort_Number  \n",
       "0          17.0              2  \n",
       "1          17.0              2  \n",
       "2          17.0              1  \n",
       "3          17.0              1  \n",
       "4           NaN              1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(f\"{raw_data_dir}/metadata.tsv\", sep=\"\\t\")\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>AssayType</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR23801859</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P027</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR23801860</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P014</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR23801861</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P043</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR23801863</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P039</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR23801864</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P028</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Run AssayType PatientID  Timepoint\n",
       "0  SRR23801859       WGS      P027         35\n",
       "1  SRR23801860       WGS      P014         14\n",
       "2  SRR23801861       WGS      P043          0\n",
       "3  SRR23801863       WGS      P039          0\n",
       "4  SRR23801864       WGS      P028          0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgs_ids = pd.read_csv(f\"{raw_data_dir}/wgs_ids.txt\", sep=\"\\t\")\n",
    "wgs_ids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Stool_Consistency</th>\n",
       "      <th>Patient_Sex</th>\n",
       "      <th>Sample_Day</th>\n",
       "      <th>Recovery_Day</th>\n",
       "      <th>Cohort_Number</th>\n",
       "      <th>Run</th>\n",
       "      <th>AssayType</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>Timepoint</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EG2537</td>\n",
       "      <td>P042</td>\n",
       "      <td>liquid</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801919</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P042</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EG2518</td>\n",
       "      <td>P043</td>\n",
       "      <td>liquid</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801861</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P043</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EG0136</td>\n",
       "      <td>P027</td>\n",
       "      <td>formed</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801959</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P027</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EG2149</td>\n",
       "      <td>P024</td>\n",
       "      <td>formed</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801920</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P024</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EG2018</td>\n",
       "      <td>P014</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>F</td>\n",
       "      <td>14</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801860</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P014</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EG1976</td>\n",
       "      <td>P014</td>\n",
       "      <td>formed</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801941</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P014</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EG1940</td>\n",
       "      <td>P071</td>\n",
       "      <td>liquid</td>\n",
       "      <td>M</td>\n",
       "      <td>27</td>\n",
       "      <td>29.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801928</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P071</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EG1939</td>\n",
       "      <td>P045</td>\n",
       "      <td>liquid</td>\n",
       "      <td>M</td>\n",
       "      <td>14</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801868</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P045</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>EG1878</td>\n",
       "      <td>P059</td>\n",
       "      <td>formed</td>\n",
       "      <td>M</td>\n",
       "      <td>22</td>\n",
       "      <td>20.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801871</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P059</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EG1830</td>\n",
       "      <td>P008</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801963</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P008</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EG0101</td>\n",
       "      <td>P054</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>F</td>\n",
       "      <td>7</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801892</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P054</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EG1599</td>\n",
       "      <td>P039</td>\n",
       "      <td>liquid</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801863</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P039</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EG1572</td>\n",
       "      <td>P026</td>\n",
       "      <td>formed</td>\n",
       "      <td>F</td>\n",
       "      <td>20</td>\n",
       "      <td>18.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801875</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P026</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EG0088</td>\n",
       "      <td>P020</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>F</td>\n",
       "      <td>27</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801935</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P020</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EG1416</td>\n",
       "      <td>P068</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>F</td>\n",
       "      <td>17</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801880</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P068</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EG1153</td>\n",
       "      <td>P016</td>\n",
       "      <td>formed</td>\n",
       "      <td>F</td>\n",
       "      <td>22</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801884</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P016</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EG1080</td>\n",
       "      <td>P036</td>\n",
       "      <td>formed</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801913</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P036</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EG1051</td>\n",
       "      <td>P028</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801864</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P028</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EG1035</td>\n",
       "      <td>P025</td>\n",
       "      <td>formed</td>\n",
       "      <td>M</td>\n",
       "      <td>70</td>\n",
       "      <td>65.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801975</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P025</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>EG0605</td>\n",
       "      <td>P007</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801865</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P007</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>EG0055</td>\n",
       "      <td>P020</td>\n",
       "      <td>liquid</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801908</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P020</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EG0447</td>\n",
       "      <td>P061</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>M</td>\n",
       "      <td>27</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801944</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P061</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>EG0381</td>\n",
       "      <td>P061</td>\n",
       "      <td>formed</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801898</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P061</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>EG0364</td>\n",
       "      <td>P049</td>\n",
       "      <td>semi-formed</td>\n",
       "      <td>M</td>\n",
       "      <td>23</td>\n",
       "      <td>25.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801977</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P049</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>EG0313</td>\n",
       "      <td>P044</td>\n",
       "      <td>liquid</td>\n",
       "      <td>F</td>\n",
       "      <td>19</td>\n",
       "      <td>23.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801999</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P044</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>EG0039</td>\n",
       "      <td>P073</td>\n",
       "      <td>formed</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801917</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P073</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>EG0282</td>\n",
       "      <td>P044</td>\n",
       "      <td>liquid</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801866</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P044</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>EG0196</td>\n",
       "      <td>P027</td>\n",
       "      <td>formed</td>\n",
       "      <td>M</td>\n",
       "      <td>35</td>\n",
       "      <td>33.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23801859</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P027</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>EG2608</td>\n",
       "      <td>P034</td>\n",
       "      <td>formed</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1</td>\n",
       "      <td>SRR23801901</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P034</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>EG0031</td>\n",
       "      <td>P021</td>\n",
       "      <td>formed</td>\n",
       "      <td>M</td>\n",
       "      <td>20</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2</td>\n",
       "      <td>SRR23802001</td>\n",
       "      <td>WGS</td>\n",
       "      <td>P021</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sample_Name Patient_ID Stool_Consistency Patient_Sex  Sample_Day  \\\n",
       "0       EG2537       P042            liquid           F           0   \n",
       "1       EG2518       P043            liquid           M           0   \n",
       "2       EG0136       P027            formed           M           0   \n",
       "3       EG2149       P024            formed           F           0   \n",
       "4       EG2018       P014       semi-formed           F          14   \n",
       "5       EG1976       P014            formed           F           0   \n",
       "6       EG1940       P071            liquid           M          27   \n",
       "7       EG1939       P045            liquid           M          14   \n",
       "8       EG1878       P059            formed           M          22   \n",
       "9       EG1830       P008       semi-formed           F           0   \n",
       "10      EG0101       P054       semi-formed           F           7   \n",
       "11      EG1599       P039            liquid           M           0   \n",
       "12      EG1572       P026            formed           F          20   \n",
       "13      EG0088       P020       semi-formed           F          27   \n",
       "14      EG1416       P068       semi-formed           F          17   \n",
       "15      EG1153       P016            formed           F          22   \n",
       "16      EG1080       P036            formed           F           0   \n",
       "17      EG1051       P028       semi-formed           F           0   \n",
       "18      EG1035       P025            formed           M          70   \n",
       "19      EG0605       P007       semi-formed           M           0   \n",
       "20      EG0055       P020            liquid           F           0   \n",
       "21      EG0447       P061       semi-formed           M          27   \n",
       "22      EG0381       P061            formed           M           0   \n",
       "23      EG0364       P049       semi-formed           M          23   \n",
       "24      EG0313       P044            liquid           F          19   \n",
       "25      EG0039       P073            formed           M           0   \n",
       "26      EG0282       P044            liquid           F           0   \n",
       "27      EG0196       P027            formed           M          35   \n",
       "28      EG2608       P034            formed           F           0   \n",
       "29      EG0031       P021            formed           M          20   \n",
       "\n",
       "    Recovery_Day  Cohort_Number          Run AssayType PatientID  Timepoint  \n",
       "0           17.0              1  SRR23801919       WGS      P042          0  \n",
       "1           17.0              1  SRR23801861       WGS      P043          0  \n",
       "2           33.0              1  SRR23801959       WGS      P027          0  \n",
       "3           37.0              1  SRR23801920       WGS      P024          0  \n",
       "4           16.0              2  SRR23801860       WGS      P014         14  \n",
       "5           16.0              1  SRR23801941       WGS      P014          0  \n",
       "6           29.0              2  SRR23801928       WGS      P071         27  \n",
       "7           11.0              2  SRR23801868       WGS      P045         14  \n",
       "8           20.0              2  SRR23801871       WGS      P059         22  \n",
       "9            NaN              1  SRR23801963       WGS      P008          0  \n",
       "10           7.0              2  SRR23801892       WGS      P054          7  \n",
       "11          26.0              1  SRR23801863       WGS      P039          0  \n",
       "12          18.0              2  SRR23801875       WGS      P026         20  \n",
       "13          28.0              2  SRR23801935       WGS      P020         27  \n",
       "14          16.0              2  SRR23801880       WGS      P068         17  \n",
       "15          24.0              2  SRR23801884       WGS      P016         22  \n",
       "16          20.0              1  SRR23801913       WGS      P036          0  \n",
       "17          21.0              1  SRR23801864       WGS      P028          0  \n",
       "18          65.0              2  SRR23801975       WGS      P025         70  \n",
       "19          20.0              1  SRR23801865       WGS      P007          0  \n",
       "20          28.0              1  SRR23801908       WGS      P020          0  \n",
       "21          25.0              2  SRR23801944       WGS      P061         27  \n",
       "22          25.0              1  SRR23801898       WGS      P061          0  \n",
       "23          25.0              2  SRR23801977       WGS      P049         23  \n",
       "24          23.0              2  SRR23801999       WGS      P044         19  \n",
       "25           NaN              1  SRR23801917       WGS      P073          0  \n",
       "26          23.0              1  SRR23801866       WGS      P044          0  \n",
       "27          33.0              2  SRR23801859       WGS      P027         35  \n",
       "28          18.0              1  SRR23801901       WGS      P034          0  \n",
       "29          24.0              2  SRR23802001       WGS      P021         20  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(meta, wgs_ids, left_on=[\"Patient_ID\", \"Sample_Day\"], right_on=[\"PatientID\", \"Timepoint\"], how=\"inner\")\n",
    "merged_df  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before the genome assembly, I have:\n",
    "\n",
    "2. Downloaded the SRA Toolkit using conda - ```conda install bioconda::sra-tools```.\n",
    "2. Extracting FASTQ Files: Use the fastq-dump utility from the SRA Toolkit to convert the SRA files into .fastq.gz format. ```fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip SRRXXXXXX``` for each SRR file!\n",
    "3. Imported the ```fastq.gz``` files into QIIME2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! fastq-dump --outdir fastq --gzip --skip-technical  --readids --read-filter pass --dumpbase --split-3 --clip SRRXXXXXX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mImported ../data/raw/manifest as PairedEndFastqManifestPhred33V2 to ../data/processed/demux-paired-end.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime tools import \\\n",
    "  --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "  --input-path $raw_data_dir/manifest \\\n",
    "  --input-format PairedEndFastqManifestPhred33V2 \\\n",
    "  --output-path $data_dir/demux-paired-end.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='moshpit'></a>\n",
    "## 2. Shotgun metagenome analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='assembly'></a>\n",
    "### 2.1 Genome assembly\n",
    "\n",
    "In this step, I begin the process of recovering metagenome-assembled genomes (MAGs) by performing genome assembly. I am using MEGAHIT that takes short DNA sequencing reads, constructs a simplified De Bruijn graph, and produces longer contiguous sequences called contigs.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31m\u001b[1mError: QIIME 2 has no plugin/command named 'assembly'.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "qiime assembly assemble-megahit \\\n",
    "    --i-seqs  ./demux-paired-end.qza \\\n",
    "    --p-presets meta-sensitive \\\n",
    "    --p-num-cpu-threads 28 \\\n",
    "    --o-contigs ./contigs.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once I’ve assembled the reads into contigs, I use QUAST to evaluate the quality of my assembly.\n",
    "\n",
    "* N50: Measures the contiguity of the assembly. It’s the length of the contig where 50% of the genome is covered by contigs of that length or longer. Higher values indicate better assembly quality.\n",
    "* L50: Indicates the number of contigs needed to cover 50% of the genome’s total length. Lower values are better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime assembly evaluate-contigs \\\n",
    "    --i-contigs ./contigs.qza \\\n",
    "    --p-threads 60 \\\n",
    "    --o-visualization ./contigs.qzv \\\n",
    "    --verbose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Press the 'q' key, Control-C, or Control-D to quit. This view may no longer be accessible or work correctly after quitting.\n",
      "Press the 'q' key, Control-C, or Control-D to quit. This view may no longer be accessible or work correctly after quitting."
     ]
    }
   ],
   "source": [
    "!qiime tools view $data_dir/contigs.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='mapping'></a>\n",
    "### 2.2 Read mapping\n",
    "Before assembling MAGs, I index the contigs from the assembly step and map the original reads to these contigs using the index. This mapping helps the contig binner group contigs from the same genome. I’ll now run the next two cells to perform contig indexing and read mapping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime assembly index-contigs \\\n",
    "    --i-contigs ./contigs.qza \\\n",
    "    --p-threads 60 \\\n",
    "    --p-seed 100 \\\n",
    "    --o-index ./contigs-index.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime assembly map-reads \\\n",
    "    --i-index ./contigs-index.qza \\\n",
    "    --i-reads demux-paired-end.qza \\\n",
    "    --p-threads 60 \\\n",
    "    --p-seed 100 \\\n",
    "    --o-alignment-map ./reads-to-contigs-aln.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='binning'></a>\n",
    "### 2.3 Contig binning\n",
    "Now I move on to contig binning, where I group contigs into bins based on their likely origin from different microbial species or strains in the community. For this, I’ll use MetaBAT 2, which assigns contigs to bins by analyzing tetranucleotide frequency and abundance (coverage) data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime moshpit bin-contigs-metabat \\\n",
    "    --i-contigs ./contigs.qza \\\n",
    "    --i-alignment-maps ./reads-to-contigs-aln.qza \\\n",
    "    --p-num-threads 60 \\\n",
    "    --p-seed 100 \\\n",
    "    --o-mags ./mags.qza \\\n",
    "    --o-contig-map ./contig-map.qza \\\n",
    "    --o-unbinned-contigs ./unbinned-contigs.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='dereplication'></a>\n",
    "### 2.4 MAG dereplication\n",
    "The previous step generated a few artifacts:  \n",
    "- `mags.qza`: the MAGs for each sample.  \n",
    "- `contig-map.qza`: a mapping of MAG IDs to the contigs they include.  \n",
    "- `unbinned-contigs.qza`: contigs that couldn’t be assigned to any MAG.  \n",
    "\n",
    "From this point, I’ll focus on `mags.qza` and I dereplicate the MAG collection, considering that the 10 samples most likely share many genomes.\n",
    "\n",
    "To dereplicate the MAGs, I:  \n",
    "1. Compute hash sketches for each genome using [sourmash](https://sourmash.readthedocs.io/en/latest/), which creates compact representations of the genomes.  \n",
    "2. Compare these sketches to generate a pairwise distance matrix.  \n",
    "3. Dereplicate the genomes by clustering similar ones using a fixed similarity threshold, keeping the longest genome in each cluster.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime sourmash compute \\\n",
    "    --i-sequence-file ./mags.qza \\\n",
    "    --p-ksizes 105 \\\n",
    "    --p-scaled 100 \\\n",
    "    --o-min-hash-signature ./mags-hashes.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime sourmash compare \\\n",
    "    --i-min-hash-signature ./mags-hashes.qza \\\n",
    "    --p-ksize 105 \\\n",
    "    --o-compare-output ./mags-dist.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime moshpit dereplicate-mags \\\n",
    "    --i-mags ./mags.qza \\\n",
    "    --i-distance-matrix ./mags-dist.qza \\\n",
    "    --p-threshold 0.7 \\\n",
    "    --o-dereplicated-mags ./mags-derep.qza \\\n",
    "    --o-feature-table ./mags-table.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='dereplication'></a>\n",
    "### 2.5 MAG Quality Analysis\n",
    "\n",
    "I performed the following quality analysis on the dereplicated set of MAGs due to computational limitations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime moshpit evaluate-busco \\\n",
    "    --i-bins ./mags-derep.qza \\\n",
    "    --i-busco-db ./busco-db.qza \\\n",
    "    --p-lineage-dataset \"bacteria_odb10\" \\\n",
    "    --p-cpu 30 \\\n",
    "    --o-results-table ./busco-results.qza \\\n",
    "    --o-visualization ./mags-derep-qc.qzv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Press the 'q' key, Control-C, or Control-D to quit. This view may no longer be accessible or work correctly after quitting.\n",
      "Press the 'q' key, Control-C, or Control-D to quit. This view may no longer be accessible or work correctly after quitting."
     ]
    }
   ],
   "source": [
    "!qiime tools view $data_dir/mags-derep-qc.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='moshpit'></a>\n",
    "## 3. MAG abundance estimation\n",
    "\n",
    "\n",
    "In this section, I will estimate MAG abundance by mapping the original reads back to the recovered MAGs. I will then count the reads mapped to each genome, normalize these counts to genome lengths, and present the results using metrics such as RPKM or TPM."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='indexing'></a>\n",
    "### 3.1 MAG indexing\n",
    "I first need to index our dereplicated MAGs. I can do it using the `index-derep-mags` action from the `q2-moshpit` plugin:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# command you want to run\n",
    "qiime assembly index-derep-mags \\\n",
    "    --i-mags ./mags-derep.qza \\\n",
    "    --p-threads 60 \\\n",
    "    --p-seed 100 \\\n",
    "    --o-index ./mags-derep-index.qza\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='mapping'></a>\n",
    "### 3.2 Read mapping\n",
    "Next, I will map the original reads to the dereplicated MAGs using the index generated in the previous step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime assembly map-reads \\\n",
    "    --i-index ./mags-derep-index.qza \\\n",
    "    --i-reads ./demux-paired-end.qza  \\\n",
    "    --p-threads 60 \\\n",
    "    --p-seed 100 \\\n",
    "    --o-alignment-map ./reads-to-mags-aln.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='feature-length'></a>\n",
    "### 3.3 MAG length calculation\n",
    "I will calculate the length of each recovered MAG to normalize read counts based on genome length, using the get-feature-lengths action."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qiime moshpit get-feature-lengths \\\n",
    "    --i-features ./mags-derep.qza \\\n",
    "    --o-lengths ./mags-derep-lengths.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='abundance'></a>\n",
    "### 3.4 Abundance estimation\n",
    "I will estimate the abundance of our MAGs using RPKM (Reads Per Kilobase per Million mapped reads) as a metric, with a minimal mapping quality of 42 to include only perfectly mapped reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "qiime moshpit estimate-mag-abundance \\\n",
    "    --i-maps ./reads-to-mags-aln.qza \\\n",
    "    --i-mag-lengths ./mags-derep-lengths.qza \\\n",
    "    --p-metric \"rpkm\" \\\n",
    "    --p-min-mapq 42 \\\n",
    "    --p-threads 60 \\\n",
    "    --o-abundances ./mags-derep-ft.qza\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='mag_classification'></a>\n",
    "### 3.5 Taxonomic classification of MAGs\n",
    "\n",
    "I will perform the taxonomic classification of the assembled genomes using Kraken 2, a popular taxonomic classifier for metagenomic data. I will use a pre-built database containing k-mer profiles for archaeal, bacterial, viral, plasmid, and human sequences from NCBI's RefSeq database to compare the analyzed genomes to references.\n",
    "\n",
    "First, I fetch the databases:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime moshpit build-kraken-db \\\n",
    "    --p-collection standard \\\n",
    "    --o-kraken2-database ./kraken_standard \\\n",
    "    --o-bracken-database ./bracken_standard \\ \n",
    "    --verbose"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I perform the classification:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime moshpit classify-kraken2 \\\n",
    "    --i-seqs ./mags-derep.qza \\\n",
    "    --i-kraken2-db ./kraken_standard \\\n",
    "    --p-threads 60 \\\n",
    "    --p-memory-mapping \\\n",
    "    --o-reports ./kraken2-reports-mags.qza \\\n",
    "    --o-hits ./kraken2-hits-mags.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I will convert the Kraken 2 report `FeatureData[Kraken2Report % Properties('mags')]`and MAG taxonomy output into a QIIME-compatible taxonomy format `FeatureData[Kraken2Output % Properties('mags')]`. The Kraken 2 report provides a tree-like representation of identified taxa, while the MAG taxonomy output lists MAGs with their corresponding identified taxa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31m\u001b[1mError: QIIME 2 has no plugin/command named 'moshpit'.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime moshpit kraken2-to-mag-features \\\n",
    "    --i-reports ./kraken2-reports-mags.qza \\\n",
    "    --i-hits ./kraken2-hits-mags.qza \\\n",
    "    --o-taxonomy ./mags-taxonomy.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Taxa identified in sample 'SRR23801859':\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium adolescentis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Eubacteriales;f__Oscillospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum;ssp__Bifidobacterium longum subsp. longum\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Eubacteriales\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus;s__Streptococcus thermophilus\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Coriobacteriia;o__Coriobacteriales;f__Coriobacteriaceae;g__Collinsella;s__Collinsella stercoris;ssp__Collinsella stercoris DSM 13279\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Coprococcus;s__Coprococcus comes\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Coprococcus;s__Coprococcus comes\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Unclassified\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium adolescentis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Coprobacillaceae;g__Catenibacterium;s__Catenibacterium sp. co_0103\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter;s__Agathobacter rectalis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Dorea\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus;s__Streptococcus thermophilus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Agathobacter;s__Agathobacter rectalis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Coprobacillaceae;g__Catenibacterium\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium adolescentis\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum\n",
      "\n",
      "Taxa identified in sample 'SRR23801860':\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Eubacteriales;f__Oscillospiraceae;g__Ruthenibacterium;s__Ruthenibacterium lactatiformans\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Eubacteriales;f__Oscillospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli;ssp__Escherichia coli ED1a\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Eubacteriales;f__Oscillospiraceae;g__Ruthenibacterium;s__Ruthenibacterium lactatiformans\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster\n",
      "\n",
      "Taxa identified in sample 'SRR23801866':\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia hansenii;ssp__Blautia hansenii DSM 20583\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia hansenii;ssp__Blautia hansenii DSM 20583\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia hansenii;ssp__Blautia hansenii DSM 20583\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Micrococcales;f__Micrococcaceae;g__Rothia;s__Rothia mucilaginosa;ssp__Rothia mucilaginosa DY-18\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Negativicutes;o__Veillonellales;f__Veillonellaceae;g__Veillonella\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Micrococcales;f__Micrococcaceae;g__Rothia;s__Rothia mucilaginosa;ssp__Rothia mucilaginosa DY-18\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia hansenii;ssp__Blautia hansenii DSM 20583\n",
      "\n",
      "Taxa identified in sample 'SRR23801898':\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Coprobacillaceae;g__Thomasclavelia;s__Thomasclavelia ramosa;ssp__Thomasclavelia ramosa DSM 1402\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Coprobacillaceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Enterococcaceae;g__Enterococcus;s__Enterococcus faecium\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Phocaeicola;s__Phocaeicola vulgatus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Coprobacillaceae;g__Thomasclavelia;s__Thomasclavelia ramosa\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Micrococcales;f__Micrococcaceae;g__Rothia;s__Rothia mucilaginosa;ssp__Rothia mucilaginosa DY-18\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Coprobacillaceae;g__Thomasclavelia;s__Thomasclavelia ramosa\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Bacteroidaceae;g__Bacteroides\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Carnobacteriaceae;g__Granulicatella;s__Granulicatella adiacens;ssp__Granulicatella adiacens ATCC 49175\n",
      "\n",
      "Taxa identified in sample 'SRR23801908':\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lacticaseibacillus;s__Lacticaseibacillus rhamnosus;ssp__Lacticaseibacillus rhamnosus LOCK908\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lacticaseibacillus;s__Lacticaseibacillus rhamnosus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lacticaseibacillus;s__Lacticaseibacillus rhamnosus\n",
      "\n",
      "Taxa identified in sample 'SRR23801935':\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum;ssp__Bifidobacterium longum subsp. longum\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Unclassified\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia pseudococcoides\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Mediterraneibacter;s__Mediterraneibacter gnavus;ssp__Mediterraneibacter gnavus ATCC 29149\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum;ssp__Bifidobacterium longum subsp. longum\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum;ssp__Bifidobacterium longum subsp. longum\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Mediterraneibacter;s__Mediterraneibacter gnavus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Lactobacillaceae;g__Lactiplantibacillus;s__Lactiplantibacillus plantarum\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Coprobacillaceae;g__Thomasclavelia;s__Thomasclavelia ramosa;ssp__Thomasclavelia ramosa DSM 1402\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Mediterraneibacter;s__Mediterraneibacter gnavus\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Tannerellaceae;g__Parabacteroides;s__Parabacteroides distasonis\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Klebsiella\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Mediterraneibacter;s__Mediterraneibacter gnavus;ssp__Mediterraneibacter gnavus ATCC 29149\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Mediterraneibacter;s__Mediterraneibacter gnavus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Micrococcales;f__Micrococcaceae;g__Rothia;s__Rothia mucilaginosa;ssp__Rothia mucilaginosa DY-18\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Rikenellaceae;g__Alistipes;s__Alistipes onderdonkii;ssp__Alistipes onderdonkii subsp. vulgaris\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia pseudococcoides\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum;ssp__Bifidobacterium longum subsp. longum\n",
      "\n",
      "Taxa identified in sample 'SRR23801941':\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Longicatena;s__Longicatena caecimuris\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Sellimonas;s__Sellimonas intestinalis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia wexlerae;ssp__Blautia wexlerae DSM 19850\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Hungatella;s__Hungatella hathewayi;ssp__Hungatella hathewayi WAL-18680\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Escherichia;s__Escherichia coli\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Eubacteriales;f__Oscillospiraceae;g__Ruthenibacterium;s__Ruthenibacterium lactatiformans\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia wexlerae;ssp__Blautia wexlerae DSM 19850\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Erysipelotrichia;o__Erysipelotrichales;f__Erysipelotrichaceae;g__Longicatena;s__Longicatena caecimuris\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Peptostreptococcales;f__Peptostreptococcaceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Peptostreptococcales;f__Peptostreptococcaceae;g__Intestinibacter;s__Intestinibacter bartlettii\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Tannerellaceae;g__Parabacteroides;s__Parabacteroides merdae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Sellimonas;s__Sellimonas intestinalis\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Coriobacteriia;o__Eggerthellales;f__Eggerthellaceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Tannerellaceae;g__Parabacteroides;s__Parabacteroides merdae\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Enterocloster;s__Enterocloster bolteae\n",
      "  d__Unclassified\n",
      "\n",
      "Taxa identified in sample 'SRR23801944':\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Roseburia;s__Roseburia intestinalis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Roseburia;s__Roseburia intestinalis\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria\n",
      "\n",
      "Taxa identified in sample 'SRR23801959':\n",
      "  d__Bacteria;k__Bacteria;p__Pseudomonadota;c__Gammaproteobacteria;o__Enterobacterales;f__Enterobacteriaceae;g__Klebsiella;s__Klebsiella pneumoniae;ssp__Klebsiella pneumoniae subsp. pneumoniae\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Unclassified\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Lachnoclostridium;s__[Clostridium] scindens\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Negativicutes;o__Veillonellales;f__Veillonellaceae;g__Dialister;s__Dialister succinatiphilus\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia wexlerae;ssp__Blautia wexlerae DSM 19850\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Sellimonas;s__Sellimonas intestinalis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Eubacteriales;f__Oscillospiraceae;g__Faecalibacterium;s__Faecalibacterium prausnitzii\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Micrococcales;f__Micrococcaceae;g__Rothia;s__Rothia mucilaginosa;ssp__Rothia mucilaginosa DY-18\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia;s__Blautia wexlerae;ssp__Blautia wexlerae DSM 19850\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Sellimonas;s__Sellimonas intestinalis\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Sellimonas;s__Sellimonas intestinalis\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Bacilli;o__Lactobacillales;f__Streptococcaceae\n",
      "  d__Unclassified\n",
      "  d__Unclassified\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes;s__Anaerostipes hadrus;ssp__Anaerostipes hadrus ATCC 29173 = JCM 17467\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes;s__Anaerostipes hadrus;ssp__Anaerostipes hadrus ATCC 29173 = JCM 17467\n",
      "\n",
      "Taxa identified in sample 'SRR23801999':\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Unclassified\n",
      "  d__Bacteria;k__Bacteria\n",
      "  d__Bacteria;k__Bacteria;p__Bacillota;c__Clostridia\n",
      "  d__Bacteria;k__Bacteria;p__Actinomycetota;c__Actinomycetes;o__Bifidobacteriales;f__Bifidobacteriaceae;g__Bifidobacterium;s__Bifidobacterium longum\n",
      "  d__Bacteria;k__Bacteria\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# read in the taxonomy\n",
    "mags_taxa = q2.Artifact.load(f\"{data_dir}/mags-taxonomy.qza\").view(pd.Series)\n",
    "# read in the presence-absence table\n",
    "mags_table = q2.Artifact.load(f\"{data_dir}/mags-table.qza\").view(pd.DataFrame)\n",
    "# replace column names with the taxa\n",
    "mags_table.columns = mags_taxa[mags_table.columns]\n",
    "\n",
    "# transpose the table\n",
    "mags_table = mags_table.T\n",
    "\n",
    "for sample in mags_table.columns:\n",
    "    print(f\"Taxa identified in sample '{sample}':\")\n",
    "    for taxon in mags_table.loc[mags_table[sample] > 0, sample].index:\n",
    "        print(f\"  {taxon}\")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Combine the previous results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\n",
      "\n",
      "Aborted!\n"
     ]
    }
   ],
   "source": [
    "! qiime taxa barplot \\\n",
    "    --i-table ./mags-derep-ft.qza \\\n",
    "    --i-taxonomy ./mags-taxonomy.qza \\\n",
    "    --o-visualization ./mags-derep-barplot.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Press the 'q' key, Control-C, or Control-D to quit. This view may no longer be accessible or work correctly after quitting.\n",
      "Press the 'q' key, Control-C, or Control-D to quit. This view may no longer be accessible or work correctly after quitting."
     ]
    }
   ],
   "source": [
    "!qiime tools view $data_dir/mags-derep-barplot.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='annotation'></a>\n",
    "## 4. Functional annotation\n",
    "To discover genes and their functions within genomes, I will use the EggNOG-mapper. Functional annotation involves several steps: I will fetch the required databases, identify gene orthologs within each MAG, annotate each ortholog to determine its associated functions, and calculate the frequency of each annotation in the original samples."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='diamond'></a>\n",
    "### 4.1 Ortholog search with Diamond\n",
    "I begin by fetching the Diamond protein database, followed by ortholog identification using the `eggnog-diamond-search` action:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime moshpit fetch-diamond-db --o-diamond-db eggnog-diamond/eggnog-diamond-db.qza --verbose >> eggnog-diamond/eggnoog-diamond-db.txt &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime moshpit eggnog-diamond-search \\\n",
    "    --i-sequences ./mags-derep.qza \\\n",
    "    --i-diamond-db ./eggnog-diamond/eggnog-diamond-db \\\n",
    "    --p-num-cpus 100 \\\n",
    "    --p-db-in-memory True \\\n",
    "    --o-eggnog-hits ./diamond-orthologs.qza \\\n",
    "    --o-table ./diamond-ft.qza\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='eggnog'></a>\n",
    "### 4.2 Ortholog annotation\n",
    "I use the EggNOG-mapper to annotate the orthologs found in the previous step. I fetch the required database and then run the `aggnog-annotate` action:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime moshpit fetch-diamond-db --o-diamond-db eggnog-diamond/eggnog-diamond-db.qza --verbose >> eggnog-diamond/eggnoog-diamond-db.txt &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime moshpit eggnog-annotate \\\n",
    "    --i-eggnog-hits ./diamond-orthologs.qza \\\n",
    "    --i-eggnog-db ./eggnog-annot-db.qza \\\n",
    "    --p-num-cpus 100 \\\n",
    "    --p-db-in-memory False \\\n",
    "    --o-ortholog-annotations ./eggnog-annotations.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='extraction'></a>\n",
    "### 4.3 Annotation extraction\n",
    "\n",
    "I will extract specific annotations from the tables generated in the previous step, which include various annotation types for each ortholog, such as CAZymes, KEGG pathways, and COG categories. I will focus on COG categories and set a maximum E-value of 0.0001 to ensure only high-quality annotations are included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!qiime moshpit extract-annotations \\\n",
    "    --i-ortholog-annotations ./eggnog-annotations.qza \\\n",
    "    --p-annotation \"cog\" \\\n",
    "    --p-max-evalue 0.00001 \\\n",
    "    --o-annotation-frequency ./cog-ft.qza\n",
    "\n",
    "!qiime moshpit multiply-tables \\\n",
    "    --i-table1 ./mags-derep-ft.qza \\\n",
    "    --i-table2 ./cog-ft.qza \\\n",
    "    --o-result-table ./mags-cog-ft.qza\n",
    "\n",
    "! qiime taxa barplot \\\n",
    "    --i-table ./mags-cog-ft.qza \\\n",
    "    --o-visualization ./cog-barplot.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='trouble'></a>\n",
    "\n",
    "## 5. Troubleshooting\n",
    "\n",
    "Unfortunately, I had an issue with the databases, generating the followign error which I couldn't solve (especially, I couldn't figure out why because I just copied the commands, the respective directories for storing databases I have create before so it cannot be the reason):\n",
    "\n",
    "(1/1) Invalid value for '--i-diamond-db': Archive does not contain a\n",
    "  correctly formatted VERSION file.\n",
    "\n",
    "\n",
    "Also, when downloading the eggnog-annot database on the latest QIIME shotgun release 2024-10, I would get a PlugInError when downloading it. Trying to solve this problem, took a long time and I couldn't find a solution... \n",
    "\n",
    "I also had way too many problems before, e.g. when  building kracken database, etc. It would be nice to have an official tutorial for shotgun sequencing, because the code from the provided notebooks is not working, at least for the newer versions of QIIME2."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2-amplicon-2024.5",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
